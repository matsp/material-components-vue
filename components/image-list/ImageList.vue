<template>
  <ul
          :class="classes"
          class="mdc-image-list"
  >
      <slot/>
  </ul>
</template>

<script>
export default {
  props: {
    column: {
      type: Number,
      validator: (value) => value >= 0 && value <= 12,
      default: 0
    },
    masonry: {
      type: Boolean,
      default: false
    },
    textProtection: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    classes () {
      let standard = {}
      if (this.column > 0 && !this.masonry) {
        standard['image-list-standard-' + this.column] = true
      }

      let masonry = {}
      if (this.column > 0 && this.masonry) {
        masonry['image-list-masonry-' + this.column] = true
        masonry['mdc-image-list--masonry'] = true
      }

      let others = {}
      others['mdc-image-list--with-text-protection'] = this.textProtection

      return { ...standard, ...masonry, ...others }
    }
  }
}
</script>
