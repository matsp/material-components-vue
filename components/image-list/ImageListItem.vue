<template>
  <li class="mdc-image-list__item">
    <div
      v-if="adjustAspectRatio"
      class="mdc-image-list__image-aspect-container"
    >
      <slot name="image" />
    </div>
    <slot
      v-else
      name="image"
    />
    <div
      v-if="$slots.default"
      class="mdc-image-list__supporting"
    >
      <span class="mdc-image-list__label">
        <slot />
      </span>
    </div>
  </li>
</template>

<script>
export default {
  props: {
    adjustAspectRatio: {
      type: Boolean,
      default: true
    }
  },
  data () {
    return {
      slotObserver: undefined
    }
  },
  mounted () {
    this.updateSlot()
    this.slotObserver = new MutationObserver(() => this.updateSlot())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
  },
  beforeDestroy () {
    this.slotObserver.disconnect()
  },
  methods: {
    updateSlot () {
      if (this.$slots.image) {
        this.$slots.image.forEach(n => {
          if (n.elm instanceof Element) n.elm.classList.add('mdc-image-list__image')
        })
      }
    }
  }
}
</script>
