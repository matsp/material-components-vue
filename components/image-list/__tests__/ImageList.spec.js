import 'mutationobserver-shim'
import { mount } from '@vue/test-utils'
import ImageList from '../ImageList.vue'
import ImageListItem from '../ImageListItem'

describe('ImageList', () => {
  it('should mount', () => {
    const wrapper = mount(ImageList)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(ImageList)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-image-list')
  })

  it('should render as standard', () => {
    const wrapper = mount(ImageList, {
      propsData: {
        column: 1
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('image-list-standard-1')
  })

  it('should render as masonry', () => {
    const wrapper = mount(ImageList, {
      propsData: {
        masonry: true,
        column: 1
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-image-list--masonry')
    expect(wrapper.classes()).toContain('image-list-masonry-1')
  })

  it('should render with text-protection', () => {
    const wrapper = mount(ImageList, {
      propsData: {
        textProtection: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-image-list--with-text-protection')
  })
})

describe('ImageListItem', () => {
  it('should mount', () => {
    const wrapper = mount(ImageListItem)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('should render with no prop', () => {
    const wrapper = mount(ImageListItem)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-image-list__item')
    expect(wrapper.vm.$data.slotObserver).toBeDefined()
  })

  it('should render without adjust aspect ratio', () => {
    const wrapper = mount(ImageListItem, {
      propsData: {
        adjustAspectRatio: false
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.mdc-image-list__image-aspect-container').exists()).toBe(false)
  })

  it('should render with img', () => {
    const wrapper = mount(ImageListItem, {
      slots: {
        image: '<img src="cat.jpg" alt="Cat">'
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.mdc-image-list__image').is('img')).toBe(true)
  })
})
