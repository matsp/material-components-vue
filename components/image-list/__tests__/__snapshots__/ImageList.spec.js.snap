// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageList should render as masonry 1`] = `<ul class="mdc-image-list image-list-masonry-1 mdc-image-list--masonry"></ul>`;

exports[`ImageList should render as standard 1`] = `<ul class="mdc-image-list image-list-standard-1"></ul>`;

exports[`ImageList should render with no prop 1`] = `<ul class="mdc-image-list"></ul>`;

exports[`ImageList should render with text-protection 1`] = `<ul class="mdc-image-list mdc-image-list--with-text-protection"></ul>`;

exports[`ImageListItem should render with img 1`] = `
<li class="mdc-image-list__item">
  <div class="mdc-image-list__image-aspect-container"><img src="cat.jpg" alt="Cat" class="mdc-image-list__image"></div>
  <!---->
</li>
`;

exports[`ImageListItem should render with no prop 1`] = `
<li class="mdc-image-list__item">
  <div class="mdc-image-list__image-aspect-container"></div>
  <!---->
</li>
`;

exports[`ImageListItem should render without adjust aspect ratio 1`] = `
<li class="mdc-image-list__item">
  <!---->
</li>
`;
